src/domain/collaboration/callout/callout.service.ts

```diff
@@ -0,0 +1,286 @@
+        `Already have an aspect with the provided display name: ${displayName}`,
+        LogContext.CONTEXT
+      );
+
+    // Not idea: get the communicationGroupID to use for the comments
+    const communicationGroupID =
+      await this.namingService.getCommunicationGroupIdForContext(callout.id);
+
+    const aspect = await this.aspectService.createAspect(
+      aspectData,
+      userID,
+      communicationGroupID
+    );
+    callout.aspects.push(aspect);
+    await this.calloutRepository.save(callout);
+    return aspect;
+  }
+
+  async createCanvas(canvasData: CreateCanvasOnCalloutInput): Promise<ICanvas> {
+    const calloutID = canvasData.calloutID;
+    const callout = await this.getCalloutOrFail(calloutID, {
+      relations: ['canvases'],
+    });
+    if (!callout.canvases)
+      throw new EntityNotInitializedException(
+        `Callout (${calloutID}) not initialised`,
+        LogContext.CONTEXT
+      );
+
+    if (canvasData.nameID && canvasData.nameID.length > 0) {
+      const nameAvailable =
+        await this.namingService.isCanvasNameIdAvailableInContext(
+          canvasData.nameID,
+          callout.id
+        );
+      if (!nameAvailable)
+        throw new ValidationException(
+          `Unable to create Canvas: the provided nameID is already taken: ${canvasData.nameID}`,
+          LogContext.CHALLENGES
+        );
+    } else {
+      canvasData.nameID = this.namingService.createNameID(
+        `${canvasData.displayName}`
+      );
+    }
+
+    const canvas = await this.canvasService.createCanvas({
+      displayName: canvasData.displayName,
+      nameID: canvasData.nameID,
+      value: canvasData.value,
+    });
+    callout.canvases.push(canvas);
+    await this.calloutRepository.save(callout);
+    return canvas;
+  }
+
+  async getCanvases(
+    callout: ICallout,
+    canvasIDs?: string[]
+  ): Promise<ICanvas[]> {
+    const calloutLoaded = await this.getCalloutOrFail(callout.id, {
+      relations: ['canvases'],
+    });
+    if (!calloutLoaded.canvases)
+      throw new EntityNotFoundException(
+        `Callout not initialised, no canvases: ${callout.id}`,
+        LogContext.CONTEXT
+      );
+
+    if (!canvasIDs) {
+      return calloutLoaded.canvases;
+    }
+    const results: ICanvas[] = [];
+    for (const canvasID of canvasIDs) {
+      const canvas = calloutLoaded.canvases.find(
+        canvas => canvas.id === canvasID
+      );
+      if (!canvas)
+        throw new EntityNotFoundException(
+          `Canvas with requested ID (${canvasID}) not located within current Callout: : ${callout.id}`,
+          LogContext.CONTEXT
+        );
+      results.push(canvas);
+    }
+    return results;
+  }
+
+  async getCanvasOnCalloutOrFail(
+    calloutID: string,
+    canvasID: string
+  ): Promise<ICanvas> {
+    const canvas = await this.canvasService.getCanvasOrFail(canvasID, {
+      relations: ['callout'],
+    });
+    const callout = (canvas as Canvas).callout;
```

Comment:
Need to talk. The pattern across the codebase is to not have relation properties in Interfaces. Callout object is inside the Canvas entity but not in Callout interface. Could change the return type to `Canvas` but that triggers multiple errors. This is a design issue we have and I'll try fixing the error after resolving all conflicts. Fun(ny) part is that `getCanvasOrFail` uses `findOne` that returns `Canvas` type but the return is cast to `ICanvas` because of the return type is defined as `Promise<ICanvas>`.

```async getCanvasOrFail(
    canvasID: string,
    options?: FindOneOptions<Canvas>
  ): Promise<ICanvas> {
    const canvas = await this.canvasRepository.findOne(
      {
        id: canvasID,
      },
      options
    );
    if (!canvas)
      throw new EntityNotFoundException(
        `Not able to locate Canvas with the specified ID: ${canvasID}`,
        LogContext.CHALLENGES
      );
    return canvas;
  }```